### PR TITLE
IDatabaseValueConverter.cs → IPropertyTypeConverter, implement in DatabaseTarget, TargetWithContext & MethodCallTargetBase

### PR DESCRIPTION
**Pull request by snakefoot**
Replaced `DatabaseTarget.ParameterValueConverter` with `ConfigurationItemFactory.ParameterTypeConverter`

Instead of doing conversion based on DbType then conversion is done by ParameterType and Format. This adds control for type conversion when using SqlDbType (or others) when also using ParameterType. 

This PR pretty much converts DbType into System.Type and does conversion based on that instead (Allowing one to override the automatic conversion by using ParameterType).

### CODE REVIEW

**Comment by snakefoot** [comment]
> I'm doubting about this change. It's difficult to do special conversion that are only for the database

What conversions are you thinking about? You can always inherit from DatabaseTarget and do your stuff.

Think one should be careful with having too many API-doors from the start. Keep things simple and easy to understand, instead of trying to predict all the crazy scenarios.

**Comment by snakefoot** [comment]
@304NotModified Have now made `GetParameterValue` virtual, so it is possible to override (Like you have decided for `CreateDatabaseParameter`)

File: src/NLog/Targets/DatabaseTarget.cs [hunk 1]
```diff
@@ -285,6 +278,13 @@ public DatabaseTarget(string name) : this()
 
         internal Type ConnectionType { get; set; }
 
+        private IParameterTypeConverter ParameterTypeConverter
+        {
+            get => _parameterTypeConverter ?? (_parameterTypeConverter = ConfigurationItemFactory.Default.ParameterTypeConverter);
+            set => _parameterTypeConverter = value;
+        }
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Targets/DatabaseTarget.cs]: I want to delay the initialization until Target is being used.

File: src/NLog/Targets/DatabaseTarget.cs [hunk 2]
```diff
@@ -285,6 +278,13 @@ public DatabaseTarget(string name) : this()
 
         internal Type ConnectionType { get; set; }
 
+        private IParameterTypeConverter ParameterTypeConverter
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Targets/DatabaseTarget.cs]: No people should replace the `ConfigurationItemFactory.Default.ParameterTypeConverter` instead. Just one should do with `ConfigurationItemFactory.Default.ValueFormatter`. One can then use Format-parameter if wanting to signal that very special conversion is needed.

File: src/NLog/Targets/DatabaseTarget.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/Targets/DatabaseTarget.cs]: But ConfigurationItemFactory.Default is initialization it already? I think it's high unlikely in non using ConfigurationItemFactory.Default?

[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/Targets/DatabaseTarget.cs]: Then we can ditch the private get/set, which is nice IMO.

File: src/NLog/Targets/MethodCallParameter.cs [hunk 3]
```diff
@@ -97,6 +98,7 @@ public MethodCallParameter(string name, Layout layout, Type type)
         /// </summary>
         /// <docgen category='Parameter Options' order='10' />
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1721:PropertyNamesShouldNotMatchGetMethods", Justification = "Backwards compatibility")]
+        [Obsolete("Use property ParameterType instead. Marked obsolete on NLog 4.6")]
```
Comment (304NotModified) [inline src/NLog/Targets/MethodCallParameter.cs]: 👍

**Comment by 304NotModified** [comment]
> What conversions are you thinking about?

ConvertToBinary? That one is removed, isn't?

File: src/NLog/Targets/TargetPropertyWithContext.cs [hunk 4]
```diff
@@ -85,6 +85,6 @@ public TargetPropertyWithContext(string name, Layout layout)
         /// Gets or sets the type of the property.
         /// </summary>
         [DefaultValue(typeof(string))]
-        public Type PropertyType { get; set; } = typeof(string);
+        public Type ParameterType { get; set; } = typeof(string);
```
Comment (304NotModified) [inline src/NLog/Targets/TargetPropertyWithContext.cs]: I think propertyType is better? It's a target with properties, and not parameters?

**Comment by 304NotModified** [comment]
Could you please tell me what code/logic is functional removed/changes? That's a bit difficult to find in the diff. Thanks!

**Comment by snakefoot** [comment]
Yes but that also said incomplete, and how many examples do you have with binary data?


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


What conversions are you thinking about?

ConvertToBinary? That one is removed, isn't?

**Comment by 304NotModified** [comment]
(PS this is the last one for NLog 4.6 (rc) I guess :) )

**Comment by snakefoot** [comment]
Instead of doing conversion based on DbType then conversion is done by ParameterType and Format. This adds control for type conversion when using SqlDbType (or others) when also using ParameterType.


It pretty much converts DbType into System.Type<http://System.Type> and does conversion based on that instead


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


Could you please tell me what code/logic is functional removed/changes? That's a bit difficult to find in the diff. Thanks!

**Comment by snakefoot** [comment]
Yes planning to this the last PR. Might want to add parameter batchsize to db target (reuse the same db-cmd-object by just updating the values) instead of creating new db-cmd for every logevent


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


(PS this is the last one for NLog 4.6 (rc) I guess :) )

File: src/NLog/Targets/TargetPropertyWithContext.cs (hunk 4 above)
Comment (snakefoot) [reply src/NLog/Targets/TargetPropertyWithContext.cs]: I want to make it consistent with the name `IParameterTypeConverter`

**Comment by snakefoot** [comment]
@304NotModified Btw. going on winter-holiday for 2 weeks, so there will be a pause in committing changes from my side.

**Comment by 304NotModified** [comment]
Have a great holiday @snakefoot !

**Comment by snakefoot** [comment]
Thank you. Hope we can get this last PR in, so we can get NLog 4.6 out soon. But now you have good time to review :)



Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


Have a great holiday @snakefoot<https://github.com/snakefoot> !

File: src/NLog/Targets/DatabaseTarget.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/Targets/DatabaseTarget.cs]: ah yes, this is for editing `ConfigurationItemFactory.Default` before writing I assume

File: src/NLog/Targets/TargetPropertyWithContext.cs (hunk 4 above)
Comment (304NotModified) [reply src/NLog/Targets/TargetPropertyWithContext.cs]: maybe IParameterTypeConverter should be IPropertyConverter ... :|

Comment (304NotModified) [reply src/NLog/Targets/TargetPropertyWithContext.cs]: or IValueConverter or something like that.

Please your opinion, I could do the renames if that's OK with you

Comment (snakefoot) [reply src/NLog/Targets/TargetPropertyWithContext.cs]: Renamed to IPropertyTypeConverter

**Comment by 304NotModified** [comment]
merged :)

**Comment by snakefoot** [comment]
Amazing. The milestone is very close now :)

**Pull request merged**